include/llvm/IR/DTUpdater.h

```diff
@@ -0,0 +1,72 @@
+//===- DTUpdater.h - DomTree/Post DomTree Updater ---------------*- C++ -*-===//
+//
+//                     The LLVM Compiler Infrastructure
+//
+// This file is distributed under the University of Illinois Open Source
+// License. See LICENSE.TXT for details.
+//
+//===----------------------------------------------------------------------===//
+//
+// This file defines the DTUpdater class, which provides a uniform way to
+// update dominator related data structures.
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef LLVM_DTUPDATER_H
+#define LLVM_DTUPDATER_H
+
+#include "llvm/Analysis/PostDominators.h"
+#include "llvm/IR/Dominators.h"
+#include "llvm/IR/Instructions.h"
+#include "llvm/Support/GenericDomTree.h"
+
+namespace llvm{
+    // Maybe needs to use template to handle those class inherit from PostDomTreeBase?
+    class DTUpdater {
+    public:
+        // updateStrategy enum class
+        enum updateStrategy : unsigned char {
+            eagerly = 0,
+            deferred = 1
+        };
+        // Constructor
+        // Because most passes now use a single strategy all over one pass
+        // So we construct DTUpdater with the strategy
+        DTUpdater(updateStrategy ups_):update_DT(false),update_PDT(false),update_lazily(isLazy(ups_)){}
+        DTUpdater(DominatorTree& DT_,updateStrategy ups_):
+                DT(&DT_),update_DT(true),update_PDT(false),update_lazily(isLazy(ups_)){}
+        DTUpdater(PostDominatorTree& PDT_,updateStrategy ups_):PDT(&PDT_),update_DT(false),update_PDT(true),update_lazily(isLazy(ups_)){}
+        DTUpdater(DominatorTree& DT_,PostDominatorTree& PDT_,updateStrategy ups_):DT(&DT_),PDT(&PDT_),update_DT(true),update_PDT(true),update_lazily(isLazy(ups_)){}
+        bool hasDomTree(){return update_DT;}
+        bool hasPostDomTree(){return update_PDT;}
+        void applyUpdates(ArrayRef<DominatorTree::UpdateType> Updates);
+        void insertEdge(BasicBlock *From, BasicBlock *To);
+        void deleteEdge(BasicBlock *From, BasicBlock *To);
+        DominatorTree& getDomTree();
+        PostDominatorTree& getPostDomTree();
+        void deleteBB(BasicBlock *DelBB);
+        bool pendingDeletedBB(BasicBlock *DelBB);
+        bool pendingDT();
+        bool pendingPDT();
+        void recalculateDT(Function &F);
+        void recalculatePDT(Function &F);
+    private:
+        bool flushDelBB();
+        DominatorTree* DT= nullptr;
+        PostDominatorTree* PDT= nullptr;
+        const bool update_DT,update_PDT,update_lazily;
+        SmallVector<DominatorTree::UpdateType, 16> PendBothTreeUpdates, PendPDTUpdates;
+        SmallPtrSet<BasicBlock *, 8> DeletedBBs;
+        bool applyUpdate(DominatorTree::UpdateKind Kind, BasicBlock *From, BasicBlock *To);
+        void flushDT();
+        void flushPDT();
+        void cleanup();
+        void movependingupdates();
```

Comment:
Take a look at the naming convention llvm uses: https://llvm.org/docs/CodingStandards.html